Support API Gateway websocket api via LambdaServer

wiowou opened the pull request:
*Issue #, if available:*

*Description of changes:*
Added a `websocket` option to the `Amazon.Lambda.AspNetCoreServer` project so that API Gateway websocket api's can also use the `LambdaServer` (DI, controllers, etc) given the user does the following:

  1. References the same lambda to handle the `RouteKey`s in API Gateway for which they have controller routes set up.
  2. Sets up controller routes as `POST` with path=`RouteKey`  

Unit tests are passing.


By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

wiowou committed:
Support API Gateway websocket api via LambdaServer

bhoradc commented:
Hi @wiowou, 

Thank you for contributing the PR. Our team will review it and get back to you as soon we can. 

Regards, 
Chaitanya

wiowou commented:
Thank you, @bhoradc !

-Behram

philasmar committed:
Merge pull request #2229 from aws/dev

Sync dev and main

GarrettBeatty committed:
update permissions (#2231) (#2232)

aws-sdk-dotnet-automation committed:
Merge branch 'dev'

aws-sdk-dotnet-automation committed:
Merge branch 'dev'

GarrettBeatty committed:
Dev (#2248)

* doc: update readme to point at changelog correctly (#2243)

* Bump tj-actions/changed-files from 47.0.0 to 47.0.1 (#2239)

Bumps [tj-actions/changed-files](https://github.com/tj-actions/changed-files) from 47.0.0 to 47.0.1.
- [Release notes](https://github.com/tj-actions/changed-files/releases)
- [Changelog](https://github.com/tj-actions/changed-files/blob/main/HISTORY.md)
- [Commits](https://github.com/tj-actions/changed-files/compare/24d32ffd492484c1d75e0c0b894501ddb9d30d62...e0021407031f5be11a464abee9a0776171c79891)

---
updated-dependencies:
- dependency-name: tj-actions/changed-files
  dependency-version: 47.0.1
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

* Bump aws-actions/configure-aws-credentials from 5.0.0 to 5.1.1 (#2238)

Bumps [aws-actions/configure-aws-credentials](https://github.com/aws-actions/configure-aws-credentials) from 5.0.0 to 5.1.1.
- [Release notes](https://github.com/aws-actions/configure-aws-credentials/releases)
- [Changelog](https://github.com/aws-actions/configure-aws-credentials/blob/main/CHANGELOG.md)
- [Commits](https://github.com/aws-actions/configure-aws-credentials/compare/a03048d87541d1d9fcf2ecf528a4a65ba9bd7838...61815dcd50bd041e203e49132bacad1fd04d2708)

---
updated-dependencies:
- dependency-name: aws-actions/configure-aws-credentials
  dependency-version: 5.1.1
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

* Add grouping for GitHub Actions updates (#2247)

Group all GitHub Actions updates into a single PR.

---------

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: Samuel <sam.r.c@hotmail.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

aws-sdk-dotnet-automation committed:
Merge branch 'dev'

philasmar committed:
Merge pull request #2258 from aws/dev

chore: Daily ASP.NET Core version update in Dockerfiles

aws-sdk-dotnet-automation committed:
Merge branch 'dev'

aws-sdk-dotnet-automation committed:
Merge branch 'dev'

aws-sdk-dotnet-automation committed:
Merge branch 'dev'

aws-sdk-dotnet-automation committed:
Merge branch 'dev'

aws-sdk-dotnet-automation committed:
Merge branch 'dev'

aws-sdk-dotnet-automation committed:
Merge branch 'dev'

aws-sdk-dotnet-automation committed:
Merge branch 'dev'

philasmar committed:
Merge pull request #2296 from aws/dev

philasmar committed:
Merge pull request #2307 from aws/dev

Update Lambda Runtime Images for NET 11 (#2302)

aws-sdk-dotnet-automation committed:
Merge branch 'dev'

aws-sdk-dotnet-automation committed:
Merge branch 'dev'

aws-sdk-dotnet-automation committed:
Merge branch 'dev'

aws-sdk-dotnet-automation committed:
Merge branch 'dev'

aws-sdk-dotnet-automation committed:
Merge branch 'dev'

philasmar committed:
Merge pull request #2334 from aws/dev

Merge dev to master

aws-sdk-dotnet-automation committed:
Merge branch 'dev'

aws-sdk-dotnet-automation committed:
Merge branch 'dev'

aws-sdk-dotnet-automation committed:
Merge branch 'dev'

aws-sdk-dotnet-automation committed:
Merge branch 'dev'

aws-sdk-dotnet-automation committed:
Merge branch 'dev'

aws-sdk-dotnet-automation committed:
Merge branch 'dev'

Commit:
pr fixes: api gateway websocket api

Commit:
pr fixes: api gateway websocket api

wiowou commented on Libraries/src/Amazon.Lambda.AspNetCoreServer/APIGatewayWebsocketApiProxyFunction.cs:
```diff
@@ -0,0 +1,70 @@
+        protected override IHeaderDictionary AddMissingRequestHeaders(APIGatewayProxyRequest apiGatewayRequest, IHeaderDictionary headers)
+        {
+            headers = base.AddMissingRequestHeaders(apiGatewayRequest, headers);
+            headers["Content-Type"] = "application/json";
```
updated

wiowou commented on Libraries/src/Amazon.Lambda.AspNetCoreServer.Hosting/ServiceCollectionExtensions.cs:
```diff
@@ -27,7 +27,12 @@ public enum LambdaEventSource
+        ApplicationLoadBalancer,
+
+        /// <summary>
+        /// API Gateway HTTP API
```
done

wiowou commented on Libraries/src/Amazon.Lambda.AspNetCoreServer/APIGatewayWebsocketApiProxyFunction{TStartup}.cs:
```diff
@@ -0,0 +1,40 @@
+namespace Amazon.Lambda.AspNetCoreServer
+{
+    /// <summary>
+    /// APIGatewayWebsocketApiV2ProxyFunction is the base class that is implemented in a ASP.NET Core Web API. The derived class implements
+    /// the Init method similar to Main function in the ASP.NET Core and provides typed Startup. The function handler for
+    /// the Lambda function will point to this base class FunctionHandlerAsync method.
+    /// </summary>
```
fixed

wiowou commented on Libraries/test/Amazon.Lambda.AspNetCoreServer.Test/TestApiGatewayWebsocketApiCalls.cs:
```diff
@@ -0,0 +1,66 @@
+
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
```
removed

wiowou commented on Libraries/test/Amazon.Lambda.AspNetCoreServer.Test/values-post-withbody-websocketapi-request.json:
```diff
@@ -0,0 +1,50 @@
+    "MessageDirection": "IN",
+    "RequestTime": "04/Sep/2024:19:59:18 +0000",
+    "RequestTimeEpoch": 1725479958896,
+    "Status": null
```
changed

wiowou commented on Libraries/src/Amazon.Lambda.AspNetCoreServer.Hosting/Internal/LambdaRuntimeSupportServer.cs:
```diff
@@ -175,4 +175,45 @@ public ApplicationLoadBalancerMinimalApi(IServiceProvider serviceProvider)
         }
     }
+
+    /// <summary>
+    /// IServer for handlying Lambda events from an API Gateway Websocket API.
+    /// </summary>
```
fixed

Commit:
fixed merge conflict. Logical changes remain

Commit:
fix lambda runtime support server

Commit:
simplify calls

GarrettBeatty committed:
Merge branch 'dev' into websocket

GarrettBeatty committed:
Fix websocket Hosting wiring and simplify MinimalApi

The websocket runtime-support server was wired to the v2
APIGatewayHttpApiV2ProxyFunction handler with v2 request/response types,
so AddAWSLambdaHosting(LambdaEventSource.WebsocketApi) never ran the
websocket-specific path/method/header overrides and dropped fields like
ConnectionId. Switch it to the v1 APIGatewayWebsocketApiProxyFunction
flow and have the websocket MinimalApi inherit from the REST one,
overriding only the three websocket virtuals. Tighten doc comments on
the new types and on ParseHttpPath/ParseHttpMethod/AddMissingRequestHeaders,
and remove the stray test .sln (the project builds via the existing
.csproj from Libraries.sln).

GarrettBeatty committed:
Add autover change file for websocket support

Records the public-API addition of APIGatewayWebsocketApiProxyFunction +
LambdaEventSource.WebsocketApi as a Minor bump on Amazon.Lambda.AspNetCoreServer
and Amazon.Lambda.AspNetCoreServer.Hosting.

philasmar commented on Libraries/src/Amazon.Lambda.AspNetCoreServer.Hosting/Internal/LambdaRuntimeSupportServer.cs:
```diff
@@ -464,4 +464,70 @@ protected override void PostMarshallItemsFeatureFeature(IItemsFeature aspNetCore
+        protected override HandlerWrapper CreateHandlerWrapper(IServiceProvider serviceProvider)
+        {
+            var handler = new APIGatewayWebsocketMinimalApi(serviceProvider);
+#pragma warning disable CA2252
```
why do we need this?

GarrettBeatty replied:
https://github.com/aws/aws-lambda-dotnet/commit/46dc95692ad198e69dcc5188cef4a0da89c5db6e#diff-dec0ad4ae2f92f72fd5cabb763659b9a222002446a2e7fb9c77120d231c18d88R363 from the response streaming pr
